[WIP] container provider: adding a link to hosted kibana

enoodle opened the pull request:
Adding a link to kibana that is hosted on the provider using a route named `logging-kibana` if that route exists.

It uses a POST request to send the bearer token to Kibana which is the same way that Openshift is sending the token (for example here [1])

![screenshot from 2016-09-05 13-30-13](https://cloud.githubusercontent.com/assets/3123328/18245580/b145a260-736e-11e6-90a2-c32de5d35e90.gif)

![launch_kibana](https://cloud.githubusercontent.com/assets/3123328/18470151/fb661884-79b3-11e6-9314-ef56ad430ba3.png)

depends on: https://github.com/fabric8io/openshift-auth-proxy/pull/11

[1] https://github.com/spadgett/origin-web-console/blob/e30aa32263beacd860469cc1bd825a16fe41060b/app/views/directives/logs/_log-viewer.html#L5-L11

simon3z commented on app/controllers/ems_container_controller.rb:
```diff
@@ -41,6 +41,21 @@ def topology_data
   end
 
+  def launch_kibana
+    logging_route = ContainerRoute.find_by_name('logging-kibana')
```
@enoodle at the very least this should be configurable in settings.

enoodle replied:
@simon3z Do we want a global setting for this?

simon3z replied:
@enoodle yes a global one should be enough to start.

enoodle replied:
@simon3z :+1: I added this.

simon3z commented:
@enoodle please add a screenshot.

simon3z commented:
@enoodle we'll discuss this during our meeting next week.

Any way to grey the button if the route is not present?

enoodle commented:
@simon3z I made it disabled when the route is not present.  I am not sure that I did it in the best possible way because I had to add a method to `ManageIQ::Providers::Openshift::ContainerManager` that I am also not sure how to access from the controller so I ended up with a duplicated method.
@alongoldboim @yaacov any advice?

simon3z commented:
FYI @portante
We'll have to put this probably in a grander scheme.

portante commented on app/controllers/ems_container_controller.rb:
```diff
@@ -41,6 +41,25 @@ def topology_data
+        page << "window.open('#{url}');"
+      end
+    else
+      add_flash("A route named 'logging-kibana' is needed to connect to kibanba.", :error)
```
s/kibanba/kibana/

portante commented on app/helpers/application_helper/toolbar/ems_container_center.rb:
```diff
@@ -96,4 +96,14 @@ class ApplicationHelper::Toolbar::EmsContainerCenter < ApplicationHelper::Toolba
+    button(
+      :ems_container_launch_kibana,
+      'product product-monitoring fa-lg',
+      N_('Open a new browser window with the Kibana hosted on this Provider. This requiers the aggregated logging to be deployed on this Proider'),
```
s/requiers/requires/

portante commented on app/controllers/ems_container_controller.rb:
```diff
@@ -41,6 +41,25 @@ def topology_data
+  end
+
+  def launch_kibana
+    logging_route = kibana_route
```
Why not just use `kibana_route`?

enoodle commented on config/settings.yml:
```diff
@@ -76,6 +76,10 @@
+:container_logging:
+  :by_provider_name:
+    :_default:
+      :kibana_route: logging-kibana
```
@simon3z what do you think of that as a solution for settings per-provider? (I can also use that in other places)
cc @alongoldboim thanks for the idea!

simon3z replied:
I think having just one global setting is good enough to start with.

WRT this strategy to store per-provider settings I don't find it particularly interesting for several reasons:

1. it's hard to maintain
2. the association by name doesn't seems reliable enough
3. we're moving in a direction where we would like all these settings to be configurable with rest-api calls

enoodle commented:
@miq-bot add_label wip
something got broken with the button, I am trying to figure it out now.

enoodle commented:
Currently simple buttons can't have text anymore untill https://github.com/ManageIQ/ui-components/pull/15 will be in a new version of ui-components in ManageIQ.
This PR is not urgent so I will wait with it until then.

simon3z commented:
@enoodle I haven't checked, is there any way to grey this button in case the `kibana` route is not available?

@dclarizio @serenamarie125 can you check the UI part? Any comment/suggestion?

@miq-bot assign enoodle

enoodle commented:
@simon3z When the button is disabled its also grayed out. Ill add screen shots (although they don't look well without this patch to the UI components)

simon3z commented:
@martinpovolny @himdel can you review?

martinpovolny commented:
Looks good to me.

Should this still be a WIP?

simon3z commented:
> Should this still be a WIP?

Probably not. cc @enoodle

enoodle commented:
@martinpovolny because without the UI components update there is no text to the button and its not clear what it does I thought to wait for the UI components new version and then merge this one. If you think its acceptable to merge in the current condition then we can remove the wip tag.

simon3z commented:
@enoodle can you rename some of the methods to be more generic? E.g. using the term `logging` or `common_logging` instead of the specific tool (`kibana`)?

enoodle commented:
@simon3z I changed the names to `common_logging`. I also added `supports :common_logging` to prevent this button from being displayed on providers that doesn't support common_logging (In case someone uses the kubernetes provider)

martinpovolny commented on app/models/manageiq/providers/container_manager.rb:
```diff
@@ -19,6 +20,10 @@ class ContainerManager < BaseManager
     has_one :container_deployment, :foreign_key => :deployed_ems_id, :inverse_of => :deployed_ems
 
+    supports :common_logging do
+      unsupported_reason_add(:common_logging, 'This provider type doesn\'t support common_logging') unless respond_to?(:common_logging_route_name)
```
wrap in `_()` for i18n?

martinpovolny commented on app/controllers/ems_container_controller.rb:
```diff
@@ -37,6 +37,23 @@ def new
+        page << "window.open('#{url}');"
+      end
+    else
+      add_flash("A route named '#{route_name}' is configured to connect to the " \
```
i18n missing here

enoodle commented:
@miq-bot remove_label wip

enoodle commented:
@miq-bot remove_label wip

enoodle commented:
This now supports authenticating to elastic search with the provider's token. The nodes button also will jump to kibana with a search specific for that node.

enoodle commented:
Now kibana searches for the node's hostname which is how it is sent from fluentd instead of fqdn.
@simon3z This patch is ready to be reviewed, PTAL
@martinpovolny I added new JS code here, I think that it should be reviewed.

simon3z commented:
@enoodle please involve also @serenamarie125 @epwinchell @dclarizio for the UX/UI side (the button that was added).

On the technical side... is this including SSO or no? (Didn't review yet)

enoodle commented:
Yes this includes sso

On Nov 2, 2016 8:28 PM, "Federico Simoncelli" <notifications@github.com>
wrote:

> @enoodle <https://github.com/enoodle> please involve also @serenamarie125
> <https://github.com/serenamarie125> @epwinchell
> <https://github.com/epwinchell> @dclarizio <https://github.com/dclarizio>
> for the UX/UI side (the button that was added).
>
> On the technical side... is this including SSO or no? (Didn't review yet)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ManageIQ/manageiq/pull/10648#issuecomment-257956575>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AC-ogCgKR74igP98m2egiQW-I0hofkbhks5q6NY5gaJpZM4JpQrM>
> .
>

simon3z commented:
> Yes this includes sso

@enoodle since this includes SSO then it must be reviewed carefully. Please walk through this with @bazulay. I am particularly worried that the secret token we hold in ManageIQ is leaked somewhere (e.g. you pass it to the client browser, etc.).

simon3z commented:
@enoodle have you reviewed this with @bazulay ?

enoodle commented:
@simon3z Not yet.

enoodle commented on app/controllers/mixins/containers_common_logging_mixin.rb:
```diff
@@ -0,0 +1,39 @@
+    route_name = ems.common_logging_route_name
+    logging_route = ContainerRoute.find_by(:name => route_name, :ems_id => ems.id)
+    if logging_route
+      user_token = SecureRandom.base64(15)
```
I was wondering if I should move this whole piece of code to the model, maybe to contianer_manager_mixin, any thoughts on that anyone?
Ideally this should be there but I also think it should be in a specific file for "common_logging" operations.

enoodle commented:
closing for https://github.com/ManageIQ/manageiq/pull/13319 and https://github.com/ManageIQ/manageiq-ui-classic/pull/36
